test/test_attn.py

```diff
@@ -0,0 +1,68 @@
+import torch
+from torch.autograd import gradcheck, Function
+from typing import Tuple
+
+
+# This file contains the custom autograd function of 'attn' for the purpose
+# of verifying that the manually derived gradients are correct
+
+class AttentionFunction(Function):
+
+    @staticmethod
+    def forward(ctx, q: torch.Tensor, k: torch.Tensor, v: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        k_t = k.t()
+        mm_output = torch.mm(q, k_t)
+        tanh_output = torch.tanh(mm_output)
+        output = torch.mm(tanh_output, v)
+
+        ctx.save_for_backward(q, k, v, mm_output, tanh_output)
+
+        return output, tanh_output
+
+    @staticmethod
+    def backward(ctx, grad_out: torch.Tensor, grad_tanh_out: torch.Tensor):
+
+        q, k, v, mm_output, tanh_output = ctx.saved_tensors
+        grad_q = grad_k = grad_v = None
+
+        if ctx.needs_input_grad[0]:  # q
+            dtanh_out = 1. - (torch.tanh(mm_output) ** 2)
```

Comment:
It would be nice to re-use computations between the different cases here.
In particular, depending on which ones you decide to implement together for your c++ implementation, you can do the same collapsing here and test all the re-use in advance (before writing them in c++).

Reply:
Good feedback, I changed the code accordingly.